File: chameneos/src/chameneos.c
/* Compile under Linux with "cc chameneos.c -o chameneos -lpthread"; run with
"./chameneos". Does not run under macOS as unnamed semaphores (using sem_init)
are not supported.

Unnamed pthread semaphores are used. The mall and each worker are threads that
synchronize with three binay semaphores:
- req_meet: for chameneos arriving at mall and requesting to meet
- sent_meet: for chameneos notifying that they provided their color
- req_other: for chameneos waiting to receive the color of their partner
Each chameneos has a local variable with the color of their partner. Chameneos
pass a pointer to that local variable together with their own color to the mall.

In C, a program terminates when the main program terminates. Here, the main
program waits for the mall thread to terminate.
*/

#include <pthread.h>
#include <stdio.h>
#include <semaphore.h>

#define Chams 100
#define Rounds 1000

typedef enum {blue, red, yellow} Color;

sem_t req_meet, sent_meet, req_other;
Color color;
Color *reply;

void *chameneos(void *colptr) {
  Color col = *((int *) colptr);
  Color othercol;
  for (;;) {
    // in forest
    sem_wait(&req_meet);
    color = col;
    reply = &othercol;
    sem_post(&sent_meet);
    // waiting to meet
    sem_wait(&req_other);
    // changing color
    if (col != othercol) col = 3 - col - othercol;
  }
}

void *mall(void *ptr) {
  int diff = 0;
  for (int r = 0; r < Chams * Rounds / 2; r++) {
    sem_wait(&sent_meet);
    Color fstcol = color;
    Color *fstreply = reply;
    sem_post(&req_meet);
    sem_wait(&sent_meet);
    Color sndcol = color;
    Color *sndreply = reply;
    sem_post(&req_meet);
    *fstreply = sndcol;
    *sndreply = fstcol;
    sem_post(&req_other);
    sem_post(&req_other);
    if (fstcol != sndcol) diff += 1;
  }
  printf("Color changes: %d\n", diff);
}

int main(int argc, char *argv[]) {
  pthread_t mid, cid[Chams];  // id's of mall, chameneos
  sem_init(&req_meet, 0, 1);  // semaphore req_meet = 1
  sem_init(&sent_meet, 0, 0); // semaphore sent_meet = 0
  sem_init(&req_other, 0, 0); // semaphore req_other = 0
  pthread_create(&mid, NULL, mall, NULL);
  for (int i = 0; i < Chams; i++) {
    int col = i % 3;
    pthread_create(&cid[i], NULL, chameneos, &col);
  }
  pthread_join(mid, NULL);
}

File: chameneos/src/chameneos.java
/* Compile under Linux and macOS with "javac chameneos.java"; run with
"java Mall".

Java monitors with signalling by notify and wait are used. Classes Chameneos and
Mall define active objects in the sense that they have methods that can be
called and each has a thread that calls another objects: each chameneos calls
method arrive of the mall, passing a pointer to itself. The mall is monitor in
the sense that at any time only one chameneos can call the method arrive. When
that method is called by the second chameneos, the mall thread is notified using
the implicit condition variable of the mall. The thread will then call back the
method meet of the two chameneos, passing the color of the other chameneos. The
chameneos can then change their color and go back to the forest. The mall thread
will also notify all further chameneos who got blocked at the call of method
arrive using the implicit condition variable of the mall. A single condition
variable is used, which means that when the second arriving chameneos notifies
the mall thread, all other waiting chameneos are notified as well.

In Java, a program terminates when all non-demon threads terminate. Here, the
chameneos are demon threads and the mall is a non-demon thread.
*/

class Chameneos extends Thread {
	final int Blue = 0, Red = 1, Yellow = 2;
	int col;
	Mall mall;
	Chameneos(int c, Mall m) {
		col = c; mall = m; setDaemon(true);
	}
	synchronized void meet(int otherCol) {
		// changing color
		if (col != otherCol) col = 3 - col - otherCol;
		notify();
	}
	synchronized public void run() {
		try {
			while (true) {
			  // in forest
	      mall.arrive(this, col);
	      // waiting to meet
			  wait();
			}
		} catch (InterruptedException e) {}
	}
}
class Mall extends Thread {
	static final int Chams = 100;
	static final int Rounds = 1000;
	Chameneos fstCham, sndCham; // invariant: sndCham != null ==> fstCham != null
	int fstCol, sndCol;
	Mall() {
		start();
	}
	synchronized void arrive(Chameneos ch, int c) throws InterruptedException {
		while (sndCham != null) wait();
		if (fstCham == null) {fstCham = ch; fstCol = c;
		} else {sndCham = ch; sndCol = c; notifyAll();
		}
	}
  synchronized public void run() {
		int diff = 0;
		for (int r = 0; r < Chams * Rounds / 2; r++) {
			while (sndCham == null)
				try {wait();} catch (InterruptedException e) {}
			fstCham.meet(sndCol); sndCham.meet(fstCol);
			if (fstCol != sndCol) diff += 1;
			fstCham = null; sndCham = null; notifyAll();
		}
		System.out.println("Color changes: " + diff);
	}
	public static void main(String args[]) {
		Mall m = new Mall();
		for (int i = 0; i < Chams; i++) new Chameneos(i % 3, m).start();
	}
}
